rust/src/builder.rs

```diff
@@ -48,6 +47,18 @@ impl<T> Builder<T> {
         }
     }
 
+    /// Get the internal byte-aligned memory buffer as a mutable slice
+    pub fn slice_mut(&self, start: usize, end: usize) -> &mut [T] {
+        unsafe {
+            slice::from_raw_parts_mut(self.data.offset(start as isize), (end - start) as usize)
+        }
+    }
+
+    /// Override the length
+    pub fn set_len(&mut self, len: usize) {
```

Comment:
So this function should semantically be marked as `unsafe`, since the user could set this to whatever value (even higher then the actual capacity).

Reply:
Isn't it better to add assert that we don't overflow the buffer and expose the function as safe?

Reply:
OK I have time to look at this today. I agree with @luben that for this method, adding assertions to make it safe is the way to go and I just pushed a PR for that.

I also think @crepererum raises good points so I have filed seperate JIRA https://issues.apache.org/jira/browse/ARROW-2417 to review the API for safety.